File: Php/app/Cashgame/Matrix/TableModel.php
namespace app\Cashgame\Matrix{

	use entities\Homegame;
	use app\Cashgame\Matrix\RowModel;
	use app\Cashgame\Matrix\ColumnHeaderModel;
	use entities\CashgameSuite;

	class TableModel{

		private $homegame;
		public $results;
		public $cashgames;
		public $suite;
		public $showYear;
		public $columnHeaderModels;
		public $rowModels;

		public function __construct(Homegame $homegame, CashgameSuite $suite){
			$this->homegame = $homegame;
			$this->results = $suite->getTotalResults();
			$this->cashgames = $suite->getCashgames();
			$this->suite = $suite;
			$this->showYear = $this->spansMultipleYears();
			$this->columnHeaderModels = $this->getHeaderModels();
			$this->rowModels = $this->getRowModels();
		}

		private function getHeaderModels(){
			$models = array();
			foreach($this->cashgames as $cashgame){
				$models[] = new ColumnHeaderModel($this->homegame, $cashgame, $this->showYear);
			}
			return $models;
		}

		private function getRowModels(){
			$models = array();
			$rank = 0;
			foreach($this->results as $result){
				$rank++;
				$models[] = new RowModel($this->homegame, $this->suite, $result, $rank);
			}
			return $models;
		}

		private function spansMultipleYears(){
			$years = array();
			foreach($this->cashgames as $cashgame){
				$year = $cashgame->getStartTime()->format('Y');
				if(!in_array($year, $years)){
					$years[] = $year;
				}
			}
			return count($years) > 1;
		}

	}

}

File: Web/Models/TableModel.cs
using System.Collections.Generic;
using System.Linq;
using Core.Classes;

namespace Web.Models{

	public class TableModel{

	    public bool ShowYear { get; set; }
	    public List<ColumnHeaderModel> ColumnHeaderModels { get; set; }
	    public List<RowModel> RowModels { get; set; }

		public TableModel (Homegame homegame, CashgameSuite suite){
			var results = suite.TotalResults;
			var cashgames = suite.Cashgames;
			ShowYear = SpansMultipleYears(cashgames);
			ColumnHeaderModels = GetHeaderModels(homegame, cashgames);
            RowModels = GetRowModels(homegame, suite, results);
		}

		private List<ColumnHeaderModel> GetHeaderModels(Homegame homegame, IEnumerable<Cashgame> cashgames)
		{
		    return cashgames.Select(cashgame => new ColumnHeaderModel(homegame, cashgame, ShowYear)).ToList();
		}

	    private List<RowModel> GetRowModels(Homegame homegame, CashgameSuite suite, IEnumerable<CashgameTotalResult> results){
            var models = new List<RowModel>();
			var rank = 0;
			foreach(var result in results){
				rank++;
				models.Add(new RowModel(homegame, suite, result, rank));
			}
			return models;
		}

		private bool SpansMultipleYears(IEnumerable<Cashgame> cashgames)
		{
		    var years = new List<int>();
			foreach(var cashgame in cashgames){
                if (cashgame.StartTime.HasValue)
                {
                    var year = cashgame.StartTime.Value.Year;
                    if (!years.Contains(year))
                    {
                        years.Add(year);
                    }
                }
			}
			return years.Count > 1;
		}

	}

}